### ISSUE TITLE
No zlib in WASI

### ISSUE THREAD

**Issue by DaAwesomeP** [python/cpython#93819]
**Bug report**
When compiling the WASI version using the docker image, it does not appear to include zlib in the build, or maybe the `run-python-wasi.sh` is not configured to allow zlib to work: `ModuleNotFoundError: No module named 'zlib'`

If this is already a known issue, then I can open a  pull to update [the docs](https://github.com/python/cpython/blob/6b330002b19d8cc3a5733b5ce030b199deb37f95/Tools/wasm/README.md).

**Your environment**

- CPython versions tested on: `main` 30610d28374f5a9698d456cebf3ae496ac01af51
- Operating system and architecture: `quay.io/tiran/cpythonbuild:emsdk3` x86_64

**Comment by DaAwesomeP** [comment]
Reading more docs, this may be really because `--enable-wasm-dynamic-linking` is not supported on WASI?

**Comment by robbertvanginkel** [comment]
I ran into this as well when trying to use the precompiled stdlib from `make wasm_stdlib` in a wasi build. This would fail as the script that produces the stdlib compresses with zlib:

https://github.com/python/cpython/blob/1de4395f62bb140563761ef5cbdf46accef3c550/Tools/wasm/wasm_assets.py#L221-L224

I managed to build my own wasi python with zlib with the following dockerfile:

```Dockerfile
# syntax=docker/dockerfile:1.4

# must be linux/amd64 as wasi-sdk only has amd64 binaries
FROM --platform=linux/amd64 debian:bullseye
RUN apt-get update \
 && apt-get install -y xz-utils git libxml2 build-essential pkg-config zlib1g-dev curl python3 \
 && rm -rf /var/lib/apt/lists/*
ADD https://github.com/python/cpython/archive/refs/tags/v3.11.1.tar.gz /src/python-v3.11.1.tar.gz
ADD https://github.com/WebAssembly/wasi-sdk/releases/download/wasi-sdk-16/wasi-sdk-16.0-linux.tar.gz /src/
ADD https://www.zlib.net/zlib-1.2.13.tar.xz /src/
ADD https://github.com/bytecodealliance/wasmtime/releases/download/v4.0.0/wasmtime-v4.0.0-x86_64-linux.tar.xz /tmp/wasmtime.tar.xz
WORKDIR /build
RUN mkdir /opt/wasi-sdk && tar -xf /src/wasi-sdk-16.0-linux.tar.gz --strip-components=1 -C /opt/wasi-sdk/ \
    	&& tar -xf /tmp/wasmtime.tar.xz --strip-components=1 -C /usr/local/bin \
	&& tar -xf /src/python-v3.11.1.tar.gz \
	&& tar -xf /src/zlib-1.2.13.tar.xz
RUN <<EOF
set -e
export PATH="/opt/wasi-sdk/bin:$PATH"
cd zlib-1.2.13/
CC=/opt/wasi-sdk/bin/clang RANLIB=/opt/wasi-sdk/bin/ranlib ./configure --prefix=
make install \
	prefix=/opt/wasi-sdk/share/wasi-sysroot \
	libdir=/opt/wasi-sdk/share/wasi-sysroot/lib/wasm32-wasi \
	pkgconfigdir=/opt/wasi-sdk/share/wasi-sysroot/lib/pkgconfig
cd ../cpython-3.11.1
./Tools/wasm/wasm_build.py wasi
/opt/wasi-sdk/bin/llvm-strip /build/cpython-3.11.1/builddir/wasi/python.wasm
cd builddir/wasi
make wasm_stdlib
EOF
```
With that, I can run an interpreter with the compressed stdlib:
```
$ docker run -it <ref to build docker file above>
root@4b26fe2dbb1b:/build# cd cpython-3.11.1/builddir/wasi/
root@4b26fe2dbb1b:/build/cpython-3.11.1/builddir/wasi# wasmtime run --mapdir /usr/local/lib::usr/local/lib python.wasm
Python 3.11.1 (main, Jan 16 2023, 20:30:30) [Clang 14.0.4 (https://github.com/llvm/llvm-project 29f1039a7285a5c3a9c353d05414 on wasi
Type "help", "copyright", "credits" or "license" for more information.
>>> 
```

The build is a bit akward since it copies zlib into the wasi-sdk sysroot, which I did because the `wasi-env` script seems to override any other place it could be discovered from:
https://github.com/python/cpython/blob/1de4395f62bb140563761ef5cbdf46accef3c550/Tools/wasm/wasi-env#L63-L64

I'd be open to contributing something to get this working by default for the wasi build, but couldn't find where the buildbot containers are defined and wanted to get some feedback on approaches before starting to modify things in the `Tools/Wasm` dir.

**Comment by brettcannon** [comment]
I believe the current buildbots are using https://github.com/tiran/ansible-wasm-buildbot to set them up.

**Comment by gpshead** [comment]
FYI this is probably the only issue blocking me making zlib required per https://discuss.python.org/t/lets-make-zlib-required-rather-than-optional-to-build-cpython/23062 and https://github.com/python/cpython/issues/91246

**Comment by kushaldas** [comment]
I can verify that static linking (of zlib) works as mentioned by @robbertvanginkel in https://github.com/python/cpython/issues/93819#issuecomment-1384541880.

**Comment by brettcannon** [comment]
https://github.com/singlestore-labs/python-wasi also builds zlib, so there might be some inspiration there.

**Comment by brettcannon** [comment]
https://discord.com/channels/453584038356058112/915046161126137856/1070023227667124365 from Singlestore Labs suggests:

> For the most part, I would modify the Makefiles to add --target=wasm32-unknown-wasi to the compiler flags. I would also add wasix to the compiler include flags and linker flags to fill the gaps between Wasi and POSIX. https://github.com/singlestore-labs/wasix

I know Christian did a bunch of work to eliminate the WASIX dependency from our builds.

**Comment by gzurl** [comment]
Folks, we are already offering a zlib build for WASI [here](https://github.com/vmware-labs/webassembly-language-runtimes/releases/tag/libs%2Fzlib%2F1.2.13%2B20230310-c46e363) as part of the [WebAssembly Language Runtimes](https://github.com/vmware-labs/webassembly-language-runtimes) project. Let us know if this fulfills your needs.

**Comment by brettcannon** [comment]
@gzurl is there something you can upstream for this?

**Comment by gzurl** [comment]
@brettcannon, yes, we can take a look at this 👍 (most likely after [CNCF Wasm Day](https://colocatedeventseu2023.sched.com/overview/type/Cloud+Native+Wasm+Day))

**Comment by brettcannon** [comment]
> will you be willing to depend on WLR releases

Unfortunately I would rather not. We try to keep CPython has dependency-free as possible.

I honestly don't know why it's turned off to begin with since I would assume zlib is easy to compile.

**Comment by brettcannon** [comment]
Thanks to @katharosada in https://github.com/python/cpython/issues/91246#issuecomment-1603957637 , we have verified that building zlib is just due to a lack of source, not an inherent issue with WebAssembly.

I'm closing this in favour of https://github.com/python/cpython/issues/91246 which would make zlib a required part of CPython and take care of the source availability problem.
